File: docs/breez-sdk/snippets/csharp/ClientSigning.cs
using Breez.Sdk.Spark;

namespace BreezSdkSnippets
{
    class ClientSigning
    {
        async Task<SignedTransferPackage> SignPackage(
            ExternalSparkSigner signer,
            UnsignedTransferPackage unsigned)
        {
            // ANCHOR: client-signing-sign-package
            TransferSignature signature;
            switch (unsigned)
            {
                case UnsignedTransferPackage.Transfer transfer:
                    // Show the user what they are approving before signing
                    var destination = transfer.target switch
                    {
                        TransferTarget.Spark spark => spark.address,
                        TransferTarget.Lightning lightning => lightning.bolt11,
                        TransferTarget.CoopExit coopExit => coopExit.address,
                        _ => throw new Exception("Unknown transfer target")
                    };
                    Console.WriteLine($"Approve sending {transfer.amountSat} sats " +
                        $"(fee {transfer.feeSat} sats) to {destination}");
                    signature = new TransferSignature.Transfer(
                        signed: await signer.PrepareTransfer(transfer.prepareTransfer)
                    );
                    break;
                case UnsignedTransferPackage.Swap swap:
                    Console.WriteLine("Approve re-shaping funds for a " +
                        $"{swap.amountSat} sat send (fee {swap.feeSat} sats)");
                    signature = new TransferSignature.Transfer(
                        signed: await signer.PrepareTransfer(swap.prepareTransfer)
                    );
                    break;
                case UnsignedTransferPackage.Token token:
                    if (token.isSwap)
                    {
                        Console.WriteLine("Approve combining token outputs for a " +
                            $"{token.tokenIdentifier} send");
                    }
                    else
                    {
                        Console.WriteLine($"Approve sending {token.amount} of token " +
                            $"{token.tokenIdentifier} (fee {token.fee})");
                    }
                    signature = new TransferSignature.Token(
                        signed: await signer.PrepareTokenTransaction(token.prepareTokenTransaction)
                    );
                    break;
                default:
                    throw new Exception("Unknown transfer package");
            }

            var signedPackage = new SignedTransferPackage(unsigned: unsigned, signature: signature);
            // ANCHOR_END: client-signing-sign-package
            return signedPackage;
        }

        async Task<Payment> SendWithClientSigning(BreezSdk sdk, ExternalSparkSigner signer)
        {
            // ANCHOR: client-signing-send
            var prepareResponse = await sdk.PrepareSendPayment(
                request: new PrepareSendPaymentRequest(
                    paymentRequest: new PaymentRequest.Input(input: "<spark address or invoice>"),
                    amount: 5_000UL,
                    tokenIdentifier: null,
                    conversionOptions: null,
                    feePolicy: null
                )
            );

            while (true)
            {
                var unsigned = await sdk.BuildUnsignedTransferPackage(
                    request: new BuildUnsignedTransferPackageRequest(
                        prepareResponse: prepareResponse,
                        options: null
                    )
                );

                // Send the package to the user, who reviews and signs it
                var signedPackage = await SignPackage(signer, unsigned);

                var response = await sdk.PublishSignedTransferPackage(
                    request: new PublishSignedTransferPackageRequest(signedPackage: signedPackage)
                );

                switch (response)
                {
                    // The wallet's funds were re-shaped first: build the payment again
                    case PublishSignedTransferPackageResponse.SwapCompleted:
                        continue;
                    case PublishSignedTransferPackageResponse.PaymentSent paymentSent:
                        return paymentSent.payment;
                }
            }
            // ANCHOR_END: client-signing-send
        }

        async Task BuildOnchainPackage(BreezSdk sdk, PrepareSendPaymentResponse prepareResponse)
        {
            // ANCHOR: client-signing-build-onchain-options
            // For Bitcoin address sends, the confirmation speed is chosen when
            // building the package: the fee depends on it
            var unsigned = await sdk.BuildUnsignedTransferPackage(
                request: new BuildUnsignedTransferPackageRequest(
                    prepareResponse: prepareResponse,
                    options: new BuildTransferPackageOptions.BitcoinAddress(
                        confirmationSpeed: OnchainConfirmationSpeed.Medium
                    )
                )
            );
            // ANCHOR_END: client-signing-build-onchain-options
            Console.WriteLine($"Unsigned package: {unsigned}");
        }

        async Task BuildBolt11Package(BreezSdk sdk, PrepareSendPaymentResponse prepareResponse)
        {
            // ANCHOR: client-signing-build-bolt11-options
            var unsigned = await sdk.BuildUnsignedTransferPackage(
                request: new BuildUnsignedTransferPackageRequest(
                    prepareResponse: prepareResponse,
                    options: new BuildTransferPackageOptions.Bolt11Invoice(
                        preferSpark: true,
                        completionTimeoutSecs: 10
                    )
                )
            );
            // ANCHOR_END: client-signing-build-bolt11-options
            Console.WriteLine($"Unsigned package: {unsigned}");
        }

        async Task<LnurlPayResponse> LnurlPayWithClientSigning(
            BreezSdk sdk,
            ExternalSparkSigner signer,
            PrepareLnurlPayResponse prepareResponse)
        {
            // ANCHOR: client-signing-lnurl-pay
            while (true)
            {
                var unsigned = await sdk.BuildUnsignedLnurlPayPackage(
                    request: new BuildUnsignedLnurlPayPackageRequest(
                        prepareResponse: prepareResponse
                    )
                );

                var signedPackage = await SignPackage(signer, unsigned);

                var response = await sdk.PublishSignedLnurlPayPackage(
                    request: new PublishSignedLnurlPayPackageRequest(signedPackage: signedPackage)
                );

                switch (response)
                {
                    case PublishSignedLnurlPayResponse.SwapCompleted:
                        continue;
                    case PublishSignedLnurlPayResponse.PaymentSent paymentSent:
                        return paymentSent.response;
                }
            }
            // ANCHOR_END: client-signing-lnurl-pay
        }
    }
}

File: docs/breez-sdk/snippets/rust/src/client_signing.rs
use std::sync::Arc;

use anyhow::Result;
use breez_sdk_spark::signer::ExternalSparkSigner;
use breez_sdk_spark::*;
use log::info;

async fn sign_package(
    signer: &Arc<dyn ExternalSparkSigner>,
    unsigned: UnsignedTransferPackage,
) -> Result<SignedTransferPackage> {
    // ANCHOR: client-signing-sign-package
    let signature = match &unsigned {
        UnsignedTransferPackage::Transfer {
            prepare_transfer,
            amount_sat,
            fee_sat,
            target,
        } => {
            // Show the user what they are approving before signing
            let destination = match target {
                TransferTarget::Spark { address, .. } => address,
                TransferTarget::Lightning { bolt11, .. } => bolt11,
                TransferTarget::CoopExit { address, .. } => address,
            };
            info!("Approve sending {amount_sat} sats (fee {fee_sat} sats) to {destination}");
            TransferSignature::Transfer {
                signed: signer.prepare_transfer(prepare_transfer.clone()).await?,
            }
        }
        UnsignedTransferPackage::Swap {
            prepare_transfer,
            amount_sat,
            fee_sat,
            ..
        } => {
            info!("Approve re-shaping funds for a {amount_sat} sat send (fee {fee_sat} sats)");
            TransferSignature::Transfer {
                signed: signer.prepare_transfer(prepare_transfer.clone()).await?,
            }
        }
        UnsignedTransferPackage::Token {
            prepare_token_transaction,
            token_identifier,
            amount,
            fee,
            is_swap,
            ..
        } => {
            if *is_swap {
                info!("Approve combining token outputs for a {token_identifier} send");
            } else {
                info!("Approve sending {amount} of token {token_identifier} (fee {fee})");
            }
            TransferSignature::Token {
                signed: signer
                    .prepare_token_transaction(prepare_token_transaction.clone())
                    .await?,
            }
        }
    };

    let signed_package = SignedTransferPackage {
        unsigned,
        signature,
    };
    // ANCHOR_END: client-signing-sign-package
    Ok(signed_package)
}

async fn send_with_client_signing(
    sdk: &BreezSdk,
    signer: &Arc<dyn ExternalSparkSigner>,
) -> Result<Payment> {
    // ANCHOR: client-signing-send
    let prepare_response = sdk
        .prepare_send_payment(PrepareSendPaymentRequest {
            payment_request: PaymentRequest::Input {
                input: "<spark address or invoice>".to_string(),
            },
            amount: Some(5_000),
            token_identifier: None,
            conversion_options: None,
            fee_policy: None,
        })
        .await?;

    loop {
        let unsigned = sdk
            .build_unsigned_transfer_package(BuildUnsignedTransferPackageRequest {
                prepare_response: prepare_response.clone(),
                options: None,
            })
            .await?;

        // Send the package to the user, who reviews and signs it
        let signed_package = sign_package(signer, unsigned).await?;

        match sdk
            .publish_signed_transfer_package(PublishSignedTransferPackageRequest { signed_package })
            .await?
        {
            // The wallet's funds were re-shaped first: build the payment again
            PublishSignedTransferPackageResponse::SwapCompleted => continue,
            PublishSignedTransferPackageResponse::PaymentSent { payment } => {
                return Ok(payment);
            }
        }
    }
    // ANCHOR_END: client-signing-send
}

async fn build_onchain_package(
    sdk: &BreezSdk,
    prepare_response: PrepareSendPaymentResponse,
) -> Result<()> {
    // ANCHOR: client-signing-build-onchain-options
    // For Bitcoin address sends, the confirmation speed is chosen when
    // building the package: the fee depends on it
    let unsigned = sdk
        .build_unsigned_transfer_package(BuildUnsignedTransferPackageRequest {
            prepare_response,
            options: Some(BuildTransferPackageOptions::BitcoinAddress {
                confirmation_speed: OnchainConfirmationSpeed::Medium,
            }),
        })
        .await?;
    // ANCHOR_END: client-signing-build-onchain-options
    info!("Unsigned package: {unsigned:?}");
    Ok(())
}

async fn build_bolt11_package(
    sdk: &BreezSdk,
    prepare_response: PrepareSendPaymentResponse,
) -> Result<()> {
    // ANCHOR: client-signing-build-bolt11-options
    let unsigned = sdk
        .build_unsigned_transfer_package(BuildUnsignedTransferPackageRequest {
            prepare_response,
            options: Some(BuildTransferPackageOptions::Bolt11Invoice {
                prefer_spark: true,
                completion_timeout_secs: Some(10),
            }),
        })
        .await?;
    // ANCHOR_END: client-signing-build-bolt11-options
    info!("Unsigned package: {unsigned:?}");
    Ok(())
}

async fn lnurl_pay_with_client_signing(
    sdk: &BreezSdk,
    signer: &Arc<dyn ExternalSparkSigner>,
    prepare_response: PrepareLnurlPayResponse,
) -> Result<LnurlPayResponse> {
    // ANCHOR: client-signing-lnurl-pay
    loop {
        let unsigned = sdk
            .build_unsigned_lnurl_pay_package(BuildUnsignedLnurlPayPackageRequest {
                prepare_response: prepare_response.clone(),
            })
            .await?;

        let signed_package = sign_package(signer, unsigned).await?;

        match sdk
            .publish_signed_lnurl_pay_package(PublishSignedLnurlPayPackageRequest {
                signed_package,
            })
            .await?
        {
            PublishSignedLnurlPayResponse::SwapCompleted => continue,
            PublishSignedLnurlPayResponse::PaymentSent { response } => {
                return Ok(response);
            }
        }
    }
    // ANCHOR_END: client-signing-lnurl-pay
}
